File: hu.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
//#include "util/chk_time.h"

//#define _MACH

#define MAX_PAIMIAN 43//9个条+9个万+9个筒+东南西北中发白春夏秋冬梅竹兰菊
#define PAIMIAN 0
#define COUNT 1

const char scheme[] = {(char)0,(char)8,(char)12,(char)14,(char)15};

struct mahjongs {
	char  mahjong[MAX_PAIMIAN][2]; //按牌面由小到大排放的麻将 [0]表示牌面，[1]表示数量
	char *idx[MAX_PAIMIAN];        //牌面到mahjong的索引
	char  count; 
};

void Add(struct mahjongs *mj,char paimian,char count) {
	mj->mahjong[mj->count][PAIMIAN] = paimian;
	mj->mahjong[mj->count][COUNT] = count;
	mj->idx[paimian] = (char*)&mj->mahjong[mj->count];
	++mj->count;
}

//移除一个刻子
int RemoveKe(struct mahjongs *mj) {
	for(int i = 0;i < mj->count; ++i) {
		if(mj->mahjong[i][COUNT] >= 3) {
			mj->mahjong[i][COUNT] -= 3;
			return 1;
		} 
	}
	return 0;
}

//移除一个顺子
int RemoveSun(struct mahjongs *mj) {
	for(int i = 0;i < mj->count; ++i) {
		if(mj->mahjong[i][COUNT] > 0) {
			char  *next = mj->idx[mj->mahjong[i][PAIMIAN] + 1];
			char  *next_next = mj->idx[mj->mahjong[i][PAIMIAN] + 2];
			if(next && next_next && next[COUNT] > 0 && next_next[COUNT] > 0){
				mj->mahjong[i][COUNT] -= 1;
				next[COUNT] -= 1;
				next_next[COUNT] -= 1;
				return 1;
			} 
		}
	}
	return 0;
}

typedef int (*rFunc)(struct mahjongs *);

int CheckScheme(char s,struct mahjongs *mj) {
	static struct mahjongs g_pais[MAX_PAIMIAN];
	rFunc f[2] = {RemoveKe,RemoveSun};
	int j = 0;
	for(int i = 0; i < mj->count; ++i) {
		if(mj->mahjong[i][COUNT] >= 2) {
			struct mahjongs *pp = &g_pais[j++];
			memset(pp,0,sizeof(*pp));			
			for(int i = 0; i < mj->count; ++i) {
				Add(pp,mj->mahjong[i][PAIMIAN],mj->mahjong[i][COUNT]);
			}
			char *jiang = pp->idx[mj->mahjong[i][PAIMIAN]];
			jiang[COUNT] -= 2;
			int ok = 1;
			for(int i = 0; i < 4; ++i) {
				int b = ((1 << i) & s) == 0 ? 0:1;
				if(!f[b](pp)) {
					ok = 0;
					break;
				}
			}
			if(ok) {
				return 1;
			}
		}
	}
	return 0;
}

int CheckHu(char p[14]) {
	char  pp[MAX_PAIMIAN][2] = {0};
	for(int i = 0; i < 14; ++i) {
		pp[p[i]][PAIMIAN] = p[i];
		pp[p[i]][COUNT] += 1;
	}
	struct mahjongs ppp = {0};
	for(int i = 0; i < MAX_PAIMIAN; ++i) {
		if(pp[i][PAIMIAN] > 0) {
			Add(&ppp,pp[i][PAIMIAN],pp[i][COUNT]);
		}
	}
	for(char i = 0; i < sizeof(scheme); ++i) {
		if(CheckScheme(scheme[i],&ppp)){
			return 1;
		}
	}
	return 0;	
}

int main() {
	char p1[14] = {1,1,1,2,2,2,3,3,3,4,4,4,5,5};
	char p2[14] = {1,2,3,4,5,6,1,2,3,4,5,6,7,7};
	char p3[14] = {1,1,1,1,2,2,2,2,3,3,3,3,4,4};
	char p4[14] = {1,1,1,2,3,4,5,6,4,5,6,7,8,9};
	char p5[14] = {1,1,1,2,3,2,3,4,3,4,5,7,8,9};
	char p6[14] = {1,1,1,4,5,6,2,3,4,4,5,6,9,9};
	char p7[14] = {1,2,3,4,1,2,3,4,5,5,7,7,9,9};

/*	uint64_t now = chk_systick();

	for(int i = 0; i < 10000; ++i) {
		CheckHu(p1);
		CheckHu(p2);
		CheckHu(p3);
		CheckHu(p4);		
		CheckHu(p5);
		CheckHu(p6);
		CheckHu(p7);		
	}

	printf("time %dms\n",chk_systick() - now);
*/

	printf("%d\n",CheckHu(p1));
	printf("%d\n",CheckHu(p2));
	printf("%d\n",CheckHu(p3));
	printf("%d\n",CheckHu(p4));
	printf("%d\n",CheckHu(p5));
	printf("%d\n",CheckHu(p6));
	printf("%d\n",CheckHu(p7));

	return 0;
}

File: hu.go
package main

import "fmt"
import "time"

const (
	MAX_PAIMIAN = 43//9个条+9个万+9个筒+东南西北中发白春夏秋冬梅竹兰菊
)

var scheme []int8 = []int8{0,8,12,14,15}
type mahjong struct {
	paimian int8
	count   int8
}

type mahjongs struct {
	m     [MAX_PAIMIAN]mahjong
	idx   [MAX_PAIMIAN]*mahjong
	count int8
}

var mahjongs_pool [MAX_PAIMIAN]mahjongs

func (this *mahjongs) Add(paimian int8,count int8) {
	this.m[this.count].paimian = paimian
	this.m[this.count].count = count
	this.idx[paimian] = &this.m[this.count]
	this.count++
}

func (this *mahjongs) RemoveKe() bool {
	for i := int8(0);i < this.count; i++ {
		if this.m[i].count >= 3 {
			this.m[i].count -= 3
			return true
		}
	}
	return false
}

func (this *mahjongs) RemoveSun() bool {
	for i := int8(0);i < this.count; i++ {
		if this.m[i].count > 0 {
			next := this.idx[this.m[i].paimian+1]
			next_next := this.idx[this.m[i].paimian+2]
			if next != nil && next_next != nil && next.count > 0 && next_next.count > 0 {
				this.m[i].count -= 1
				next.count -= 1
				next_next.count -= 1
				return true
			}

		}
	}
	return false
}

func (this *mahjongs) Reset() {
	this.count = 0
	for i := 0;i < MAX_PAIMIAN; i++ {
		this.m[i].paimian = 0
		this.m[i].count = 0
		this.idx[i] = nil
	}
}

func CheckScheme(s int8,mj *mahjongs) bool {

	f := func (i int8,mj *mahjongs) bool {
		if i == 0 {
			return mj.RemoveKe()
		} else {
			return mj.RemoveSun()
		}
	}

	j := 0
	for i := int8(0);i < mj.count; i++ {
		if mj.m[i].count >= 2 {
			pp := &mahjongs_pool[j]
			j++
			pp.Reset()
			for i := int8(0);i < mj.count; i++ {
				pp.Add(mj.m[i].paimian,mj.m[i].count)
			}
			jiang := pp.idx[mj.m[i].paimian]
			jiang.count -= 2
			ok := true
			for i := uint8(0); i < 4; i++ {
				if !f(int8((1 << i) & s),pp) {
					ok = false
					break
				}
			}
			if ok {
				return true
			}
		}
	}
	return false
}

func CheckHu(p [14]int8) bool {
	var pp [MAX_PAIMIAN]mahjong
	for i := 0;i < 14; i++ {
		pp[p[i]].paimian = p[i]
		pp[p[i]].count++
	}

	var ppp mahjongs
	for i := 0; i < MAX_PAIMIAN; i++ {
		if pp[i].paimian > 0 {
			ppp.Add(pp[i].paimian,pp[i].count)
		}
	}
	for i := 0 ; i < len(scheme); i++ {
		if CheckScheme(scheme[i],&ppp) {
			return true
		}
	}
	return false
}

func main() {
	p1 := [14]int8{1,1,1,2,2,2,3,3,3,4,4,4,5,5}
	p2 := [14]int8{1,2,3,4,5,6,1,2,3,4,5,6,7,7}
	p3 := [14]int8{1,1,1,1,2,2,2,2,3,3,3,3,4,4}
	p4 := [14]int8{1,1,1,2,3,4,5,6,4,5,6,7,8,9}
	p5 := [14]int8{1,1,1,2,3,2,3,4,3,4,5,7,8,9}
	p6 := [14]int8{1,1,1,4,5,6,2,3,4,4,5,6,9,9}
	p7 := [14]int8{1,2,3,4,1,2,3,4,5,5,7,7,9,9}


	now := time.Now().UnixNano()/int64(time.Millisecond)
	count := 0

	for i := 0; i < 10000; i++ {
		CheckHu(p1)
		CheckHu(p2)
		CheckHu(p3)
		CheckHu(p4)
		CheckHu(p5)
		CheckHu(p6)
		CheckHu(p7)
		count += 7
	}

	elapse := time.Now().UnixNano()/int64(time.Millisecond) - now

	fmt.Printf("count:%d time:%dms %d per second\n",count,elapse,int64(count*1000)/elapse)

	fmt.Printf("%b\n",CheckHu(p1))
	fmt.Printf("%b\n",CheckHu(p2))
	fmt.Printf("%b\n",CheckHu(p3))
	fmt.Printf("%b\n",CheckHu(p4))
	fmt.Printf("%b\n",CheckHu(p5))
	fmt.Printf("%b\n",CheckHu(p6))
	fmt.Printf("%b\n",CheckHu(p7))

}



	
